### PR TITLE
feat: update to go-libp2p-core 0.7.0 interface changes

### PR DESCRIPTION
**Pull request by Stebalien**
This patch updates go-libp2p for the stream interface changes in go-libp2p-core 0.7.0. This is a _significant_ breaking change to streams and all users should read https://github.com/libp2p/go-libp2p-core/releases/tag/v0.7.0. In practice, this change should remove a significant footgun.

TL;DR:

* `Stream.Close` now behaves like `net.TCPConn.Close`.
* There is a new `Stream.CloseWrite` (send an EOF) and `Stream.CloseRead` (close for reading), behaving like their counterparts in `net.TCPConn`.

Dependencies:

* https://github.com/multiformats/go-multistream/pull/52

Bubbling:

* https://github.com/libp2p/go-libp2p-pubsub/pull/385
* https://github.com/ipfs/go-bitswap/pull/445
* https://github.com/libp2p/go-libp2p-kad-dht/pull/693
* https://github.com/libp2p/go-libp2p-gostream/pull/60
* https://github.com/ipfs/go-graphsync/pull/116
* https://github.com/libp2p/go-libp2p-pubsub-router/pull/81
* https://github.com/ipfs/go-ipfs/pull/7747

### CODE REVIEW

**Review by vyzo — COMMENTED** [review]
changes are nice and simple and LGTM, but let's hold a bit to figure out our upgrade path for pubsub/dht/lotus/ipfs etc.

**Comment by Stebalien** [comment]
Blocked on fixing an issue with multistream.

File: p2p/net/mock/mock_stream.go [hunk 1]
```diff
@@ -121,6 +121,15 @@ func (s *stream) Close() error {
 	return nil
 }
 
+func (s *stream) CloseRead() error {
+	return s.read.CloseWithError(ErrClosed)
+}
+
+func (s *stream) Close() error {
+	s.CloseRead()
```
Comment (aschmahmann) [inline p2p/net/mock/mock_stream.go:129]: do we need `_ = ` to satisfy the linter?

File: p2p/protocol/identify/id_delta.go [hunk 2]
```diff
@@ -26,7 +25,7 @@ func (ids *IDService) deltaHandler(s network.Stream) {
 		return
 	}
 
-	defer helpers.FullClose(s)
+	defer s.Close()
```
Comment (aschmahmann) [inline p2p/protocol/identify/id_delta.go:28]: what qualifies as requiring a reset? e.g. does failing to consume the delta warrant a reset?

File: p2p/net/mock/mock_stream.go [hunk 3]
```diff
@@ -268,20 +277,19 @@ func (s *stream) transport() {
...
-				s.resetWith(err)
+				s.cancelWrite(err)
 				return
 			}
 		}
 	}
 }
 
-func (s *stream) resetWith(err error) {
+func (s *stream) cancelWrite(err error) {
```
Comment (aschmahmann) [inline p2p/net/mock/mock_stream.go:292]: Are there any test changes downstream that would be effected by this change?

**Review by aschmahmann — APPROVED** [review]
In ping.go, should PingHandler and Ping be utilizing Closes instead of Resets now?

Otherwise LGTM

File: p2p/net/mock/mock_stream.go (hunk 1 above)
Comment (Stebalien) [reply p2p/net/mock/mock_stream.go:129]: We don't have that linter here, but it can't hurt.

File: p2p/net/mock/mock_stream.go (hunk 3 above)
Comment (Stebalien) [reply p2p/net/mock/mock_stream.go:292]: I believe I needed this to fix an issue in the bitswap tests.

File: p2p/protocol/identify/id_delta.go (hunk 2 above)
Comment (Stebalien) [reply p2p/protocol/identify/id_delta.go:28]: Yes. Although the other end likely isn't listening for an error.

**Pull request merged**